DOCSP-41134: Retrieve Data (#6)
mongodb/docs-kotlin-sync: source/read/retrieve.txt

.. _kotlin-sync-retrieve:

=============
Retrieve Data
=============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code examples, read, search, cursor

Overview
--------

In this guide, you can learn how to use the {+driver-long+} to retrieve data from a
MongoDB collection by using read operations. You can call the ``find()`` method to
retrieve documents that match a set of criteria.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

.. _kotlin-sync-retrieve-find:

Find Documents
--------------

> The {+driver-short+}'s ``find()`` method retrieves documents from a collection. This

Reviewer: S: remove driver reference; redundant

```suggestion
The ``find()`` method retrieves documents from a collection. This
```

> method takes a **query filter** and returns all matching documents. A query filter is an
> object that specifies the documents you want to retrieve in your query.

Reviewer: ```suggestion
method takes a **query filter** and returns all matching documents. A query filter is an
document that specifies the criteria that the driver uses to match documents from the collection.
```

.. To learn more about query filters, see :ref:`kotlin-sync-specify-query`.

Find Documents Example
~~~~~~~~~~~~~~~~~~~~~~

> The following example uses the ``find()`` method to find all documents where the
> ``cuisine`` field has the value ``Spanish``:

Reviewer: S: format string values

```suggestion
The following example uses the ``find()`` method to find all documents in which the
the value of the ``cuisine`` field is ``"Spanish"``:
```

.. code-block:: kotlin
   :copyable: true

>    val results = collection.find(eq("cuisine", "Spanish"))

Reviewer: S: move code examples into a file

> The ``find()`` operation in the preceding example returns a ``FindIterable`` object,
> which can be iterated over by using the ``forEach()`` method, as shown in the following
> example:

Reviewer: ```suggestion
The ``find()`` operation in the preceding example returns a ``FindIterable`` object,
which you can iterate through by using the ``forEach()`` method, as shown in the following
example:
```

.. code-block:: kotlin
   :copyable: true

   val results = collection.find(eq("cuisine", "Spanish"))
   results.forEach { result ->
     println(result)
   }

To find all documents in a collection, pass an empty filter to the ``find()`` method:

.. code-block:: kotlin
    :copyable: true

>     val allRestaurants = collection.find()

Reviewer: S: move this section to a subsection or move into a note

Modify Find Behavior
~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``find()`` method by chaining methods to
the ``find()`` method call.

> The following example uses additional methods to limit the number of documents returned

Reviewer: ```suggestion
The following example chains the ``limit()`` and ``maxTime()`` methods to limit the number of documents returned
```

Reviewer: In general, please resolve vale issues before sending PR for review!

Author: Good callout, that's on me for missing that

> by the query to 10 and set a maximum execution time of 10,000 milliseconds on the
> operation:

Reviewer: ```suggestion
by the query to ``10`` and set a maximum execution time of ``10000`` milliseconds on the
operation:
```

.. code-block:: kotlin

   val results = collection.find(eq("cuisine", "Spanish")).limit(10).maxTime(10000)

For a full list of methods that modify the behavior of ``find()``, see the `API documentation <{+api+}/mongodb-driver-kotlin-sync/com.mongodb.kotlin.client/-find-iterable/index.html>`__

> for the ``FindIterable`` class.

Reviewer: S: can we include this list as a table here to include at least some of the other important find options? Open to discussing

Author: Will grab a selection of important-seeming ones for this page!

The ``find()`` operation in the preceding example returns a ``FindIterable`` object,
which you can iterate through by using the ``forEach()`` method, as shown in the following
example:

.. literalinclude:: /includes/read/retrieve.kt
   :start-after: start-find-iterate
   :end-before: end-find-iterate
   :language: kotlin
   :copyable:
   :dedent:

.. note:: Find All Documents
 
   To find all documents in a collection, pass an empty filter to the ``find()`` method:

   .. literalinclude:: /includes/read/retrieve.kt
      :start-after: start-find-all
      :end-before: end-find-all
      :language: kotlin
      :copyable:
      :dedent:

Modify Find Behavior
~~~~~~~~~~~~~~~~~~~~

You can modify the behavior of the ``find()`` method by chaining methods to
the ``find()`` method call. The following table describes commonly used methods used for
modifying queries:

.. list-table::
   :widths: 30 70
   :header-rows: 1

   * - Method
     - Description

   * - ``batchSize()`` 
     - | Limits the number of documents to return per batch. To learn more about
         batch size, see :manual:`cursor.batchSize() </reference/method/cursor.batchSize/>`
         in the MongoDB Server manual.

>    * - ``collation()`` 
>      - | An instance of the ``Collation`` class that sets the collation options. 

Reviewer: S: change all descriptions to be the same part of speech
```suggestion
   * - ``collation()`` 
     - | Sets the collation options for the query.
```

>    * - ``comment()`` 
>      - | A string to attach to the query. This can help you trace and interpret the
>          operation in the server logs and in profile data. To learn more about query comments, 
>          see :manual:`$comment </reference/operator/query/comment/>` in the MongoDB Server
>          manual.

Reviewer: ```suggestion
   * - ``comment()`` 
     - | Specifies a string to attach to the query. This can help you trace and interpret the
         operation in the server logs and in profile data. To learn more about query comments, 
         see :manual:`$comment </reference/operator/query/comment/>` in the MongoDB Server
         manual.
```

Reviewer: and so on

   
   * - ``hint()`` 
     - | The index to use for the query.

   * - ``limit()``
     - | Limits the number of documents to be returned from the query.

   * - ``maxTime()`` 
     - | The maximum execution time on the server for this operation.

   * - ``skip()``
     - | The number of documents to skip.

   * - ``sort()``
     - | The sort criteria to apply to the query.

The following example chains the ``limit()`` and ``maxTime()`` methods to limit the
number of documents returned by the query to ``10`` and set a maximum execution time of
``10000`` milliseconds on the operation:

.. literalinclude:: /includes/read/retrieve.kt
   :start-after: start-modified-find
   :end-before: end-modified-find
   :language: kotlin
   :copyable:
   :dedent:

For a full list of methods that modify the behavior of ``find()``, see the `API documentation <{+api+}/mongodb-driver-kotlin-sync/com.mongodb.kotlin.client/-find-iterable/index.html>`__
for the ``FindIterable`` class.

Additional Information
----------------------

.. TODO: To learn more about query filters, see :ref:`kotlin-sync-specify-query`.

> For runnable code examples of retrieving documents with the {+driver-short+}, see
> :ref:`kotlin-sync-read`.

Reviewer: ```suggestion
To view runnable code examples that retrieve documents by using the {+driver-short+}, see
:ref:`kotlin-sync-read`.
```

.. _kotlin-sync-retrieve:

=============
Retrieve Data
=============

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: code examples, read, search, cursor

Overview
--------

In this guide, you can learn how to use the {+driver-short+} to retrieve data from a
MongoDB collection by using read operations. You can call the ``find()`` method to
retrieve documents that match a set of criteria specified in a query filter.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``restaurants`` collection in the ``sample_restaurants``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

> This data is modeled with the following Kotlin data class:

Reviewer: S: use `language` source constant where applicable
```suggestion
The documents in this collection are modeled by the following {+language+} data class:
```
